### DISCUSSION TITLE
Securing Dashboard Apps?

### DISCUSSION THREAD

**Discussion by nickmonad** [chatwoot/chatwoot#5878] [Q&A] [answered]
We started to use the dashboard app functionality to integrate with other tooling APIs we use for customer orders, shipping information, etc.

Everything works great, but we have a bit of a concern security wise. We would ideally like to keep our hosted dashboard app accessible by support agents and other employees only, by making the domain accessible via a private network, the same where our self-hosted chatwoot instance is hosted. (For example, the kubernetes cluster DNS)

Unfortunately, chatwoot seems to use an embedded `iframe` element to render the dashboard app, which forces the browser to make the request to the app, instead of the chatwoot server instance itself. This means the support agents must either be on the same VPN as the chatwoot server instance, or we have to make the dashboard app publicly available in our DNS records. This isn't really an option for us, as we have an endpoint in that dashboard app that allows searching for customer orders via email, and we don't want that discovered and abused in the wild.

How are people thinking about security for dashboard apps? If the customer information is provided to the dashboard app from chatwoot, it must be doing _some_ kind of API request(s) against backing customer data stores, which means that either:

1. no security - dashboard apps are public and the endpoints used to search for data based on context provided by chatwoot is completely in the clear

Or,

2. dashboard apps should have their own authentication and force support agents to login (kind of a pain, as it means we have to roll our own auth there)

I personally am not a huge fan of either of these approaches (1 is obvious why not).

This could be resolved if chatwoot _itself_ made the HTTP request and just rendered out the dashboard app internally, or had a different model entirely that could support that, but that doesn't seem likely in the near future (unless the team is working on this issue already).

Any thoughts / suggestions? Thanks!

**Comment by pranavrajs** [comment, marked as answer]
@nickmonad Thanks for starting the discussion here. 

Does the basic auth work here? You can embed the application with a basic auth so that even though it is available in the public network it would still ask for the user name and password, whereas on the dashboard app you can pass this in the URL params.

Let me know what you think.

**Reply by nickmonad** [reply]
Hey @pranavrajs thanks for the reply.

Just to be sure, you mean passing username / password as part of the URL query params when configuring the dashboard app in chatwoot? Or is there an actual HTTP basic auth setting I'm missing? I just see an option for dashboard app name and URL

Thanks!

**Reply by nickmonad** [reply]
(We're running chatwoot v2.10.0)

**Reply by nickmonad** [reply]
It seems like this is the best bet for now. Thanks!

**Reply by pranavrajs** [reply]
Missed this, yes this is the best option. We will consider adding headers in future iterations.

**Reply by nickmonad** [reply]
Thanks! I think having the option for the chatwoot server instance itself to make the HTTP request and render the HTML back out to the browser might be ultimately a more secure solution, since then the dashboard app just needs to be resolvable on the same internal network as the server. (Again, thinking about a kubernetes bases deployment where the dashboard app would just use cluster DNS)

If that's not feasible, or has it's own security risks involved I'm not aware of, perhaps the chatwoot server could make an API request to an internally resolvable service that returns JSON, and the dashboard app uses an HTML template (based on jinja or something similar) uploaded as part of the configuration step. Then the server just maps that JSON response into the template, to render back out to the user.

In any case, thanks for all the work on chatwoot!

**Reply by MichelN89** [reply]
Hi @nickmonad, I just came across this message and I'm trying to get my head around the same issue. Have you found a good solution yet?

**Reply by nickmonad** [reply]
The best we came up with was using long, random sub-paths that our chatwoot app server responds too, as well as a long, random "api key" passed via query parameters (you could argue only one of these is necessary, especially since we access the dashboard app over HTTPS and all paths and query parameters are encrypted over-the-wire, but we just haven't changed it since)

So for example, 

```
https://our.chatwoot.app.com/abbe9e3287af9d6adbd74f51a83d4015ed92ec4aa6b20eff818b0ceabf6e6bdd?key=96a74df505dceeeeee7b392a0e265fd7
```

gets configured as the dashboard app URL in the chatwoot admin settings. `https://our.chatwoot.app.com/` just returns a 404.

The server reads these values from our secrets management tool at runtime and configures them as part of the web framework routing. Rotate them as desired.

**Reply by nickmonad** [reply]
This also works for "sub-apps", if you only want to host a single dashboard app server, just do something like,

```
https://our.chatwoot.app/<random value>/app1/
https://our.chatwoot.app/<random value>/app2/
```

**Comment by donnyxray** [comment]
Are there any plans to improve this? IMHO both basic auth and using obscure URLs are not very sercure.

Chatwoot should do the fetching server-side (allowing simple IP filtering by the app) and sign the request (for optional full authentication).

**Comment by metatick** [comment]
I'd like to see some options for security here too - I'm not super comfortable with relying purely on an obscure URL for security!

Personally I don't think there's much value in request being handled server-side, as this would introduce a lot of complexity for little gain. For example:
How would the dashboard App render? Running the dashboard app in the same origin as chatwoot is a potential security issue if agents log in to multiple chatwoot accounts. 
How should it access backend services? Do they all need to be proxied too?
CORS/cookies/etc.. would need special handling
etc...

Besides that - the same level of security could, roughly, be achieved by either requiring agents to connect to a VPN so they're on the same network as the dashboard App, or by using client certificates.

But, none of that really solves the underlying issue - authenticating the agent!

I can think of two reasonable solutions that could work well:

1) Allow a simple localStorage key/value proxy via postMessage, preferably with a simple library to wrap the request/response, so we can perform authentication within the iframe and store/retrieve a JWT token (or similar) to avoid authentication being required frequently due to third party cookie/localStorage issues
This would allow something like the following within dashboard apps:
```
async function getToken(){
    let token = await chatwootStorageApi.getItem('token');
    if(!token){
       token = await functionThatShowsLoginPromptAndReturnsJWTTokenAfterSuccessfulAuthentication();
       chatwootStorageApi.setItem('token', token);
    }
    return token;
}

async function getPrivateCustomerDataFromApi(){
    const token = await getToken();
   // get/display data using JWT token
}
````

2) Have an option to allow passing a JWT token with agent information to the iframe as a query parameter, eg. https://the-dashboard-app.com?token=JWT_TOKEN_WITH_AGENT_INFORMATION . This would allow dashboard apps to handle authentication server-side using the JWT token, and either just present the information, or pass down some sort of credential for the client to use for future requests, etc...

I think both of these would be fairly straight forward... #1 should be pretty quick and easy, as it's only client-side javascript. #2 would probably be more useful though, as it would allow for authentication of the agent without requiring them to login again.

Having both would be awesome :)

Kind Regards,
Michael Thomas
